src/test/java/org/cactoos/io/ReaderOfTest.java

```diff
@@ -40,18 +41,317 @@
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
+
+    @Test
+    public void readsInputWithCharset() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't read content",
+            new InputOf(
+                new ReaderOf(
+                    new InputOf("Hello, товарищ!"),
+                    StandardCharsets.UTF_8
+                )
+            ),
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
+
+    @Test
+    public void readsInputWithCharsetByName() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't read content",
+            new InputOf(
+                new ReaderOf(
+                    new InputOf("Hello, товарищ!"),
+                    StandardCharsets.UTF_8.name()
+                )
+            ),
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
+
+    @Test
+    public void readsInputWithCharsetDecoder() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't read content",
+            new InputOf(
+                new ReaderOf(
+                    new InputOf("Hello, товарищ!"),
+                    StandardCharsets.UTF_8.newDecoder()
+                )
+            ),
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
+
+    @Test
+    public void readsInputStream() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't read content",
+            new InputOf(
+                new ReaderOf(
+                    new InputStreamOf("Hello, товарищ!")
+                )
+            ),
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
+
+    @Test
+    public void readsInputStreamWithCharset() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't read content",
+            new InputOf(
+                new ReaderOf(
+                    new InputStreamOf("Hello, товарищ!"),
+                    StandardCharsets.UTF_8
+                )
+            ),
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
+
+    @Test
+    public void readsInputStreamWithCharsetByName() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't read content",
+            new InputOf(
+                new ReaderOf(
+                    new InputStreamOf("Hello, товарищ!"),
+                    StandardCharsets.UTF_8.name()
+                )
+            ),
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
+
+    @Test
+    public void readsInputStreamWithCharsetDecoder() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't read content",
+            new InputOf(
+                new ReaderOf(
+                    new InputStreamOf("Hello, товарищ!"),
+                    StandardCharsets.UTF_8.newDecoder()
+                )
+            ),
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
```

Comment:
@proshin-roman Can you add a few tests with empty and null content? What happens if ReaderOf tries to read from an empty/null source?

Reply:
I've added tests for null and empty string.